src/channeloutput/generic_spi.cpp

```diff
@@ -0,0 +1,195 @@
+/*
+ *   Generic SPI handler for Falcon Player (FPP)
+ *
+ *   Copyright (C) 2013-2018 the Falcon Player Developers
+ *      Initial development by:
+ *      - David Pitts (dpitts)
+ *      - Tony Mace (MyKroFt)
+ *      - Mathew Mrosko (Materdaddy)
+ *      - Chris Pinkham (CaptainMurdoch)
+ *      For additional credits and developers, see credits.php.
+ *
+ *   The Falcon Player (FPP) is free software; you can redistribute it
+ *   and/or modify it under the terms of the GNU General Public License
+ *   as published by the Free Software Foundation; either version 2 of
+ *   the License, or (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <thread>
+
+#include "common.h"
+#include "log.h"
+#include "settings.h"
+
+#include "generic_spi.h"
+
+#define MAX_CHANNELS 16777215
+
+#ifdef PLATFORM_PI
+#	define MIN_SPI_SPEED_HZ 32000
+#	define MAX_SPI_SPEED_HZ 125000000
+#elif PLATFORM_BBB
+//TODO need to confirm these
+#	define MIN_SPI_SPEED_HZ 1500
+#	define MAX_SPI_SPEED_HZ 48000000
+#else
+//not sure what other platforms could be used, using safe values here
+#	define MIN_SPI_SPEED_HZ 50000
+#	define MAX_SPI_SPEED_HZ 100000
+#endif
+
+
+
+extern "C" {
+    GenericSPIOutput *createGenericSPIOutput(unsigned int startChannel,
+                                           unsigned int channelCount) {
+        return new GenericSPIOutput(startChannel, channelCount);
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////
+
+/*
+ *
+ */
+GenericSPIOutput::GenericSPIOutput(unsigned int startChannel, unsigned int channelCount)
+  : ThreadedChannelOutputBase(startChannel, channelCount),
+	m_port(-1),
+	m_speed_hz(0),
+    m_spi(nullptr)
+{
+	LogDebug(VB_CHANNELOUT, "GenericSPIOutput::GenericSPIOutput(%u, %u)\n",
+		startChannel, channelCount);
+
+	m_maxChannels = MAX_CHANNELS;
+}
+
+/*
+ *
+ */
+GenericSPIOutput::~GenericSPIOutput()
+{
+	LogDebug(VB_CHANNELOUT, "GenericSPIOutput::~GenericSPIOutput()\n");
+    if (m_spi) {
+        delete m_spi;
+    }
+}
+
+/*
+ *
+ */
+
+int GenericSPIOutput::Init(Json::Value config) {
+    char configStr[2048];
+    ConvertToCSV(config, configStr);
+    return Init(configStr);
+}
+
```

Comment:
Just a note about this:    ideally for "new" code, you would use the JSON directly for configuring the output and not convert to CSV.   The CSV stuff is legacy and the ConvertToCSV method is there to make migrating the outputs from 1.x quicker/easier.   Newer outputs should just use the JSON.     (and any help updating old code would be appreciated)

Reply:
Got it.  I'll update some of these modules to use JSON.